File: sem1/hw5/hw-5.1/hw-5.1/List.cpp
#include <stdio.h>
#include "List.h"

struct Node;

struct List
{
	Node *head = nullptr;
	int length = 0;
};

struct Node
{
	int data = 0;
	int position = 0;
	Node *next = nullptr;
};

List *createList()
{
	return new List;
}

bool isEmpty(List *list)
{
	return list->head == nullptr;
}

void addNode(List *list, int data)
{
	auto newNode = new Node{ data, list->length + 1, nullptr };
	if (isEmpty(list))
	{
		list->head = newNode;
		++list->length;
		return;
	}
	Node *nextNode = list->head;
	Node *prevNode = nullptr;
	while (nextNode != nullptr && newNode->data > nextNode->data)
	{
		++nextNode->position;
		--newNode->position;
		prevNode = nextNode;
		nextNode = prevNode->next;
	}
	if (nextNode != list->head)
	{
		newNode->next = nextNode;
		prevNode->next = newNode;
	}
	else
	{
		newNode->next = list->head;
		list->head = newNode;
	}
	++list->length;
}

void deleteNode(List *list, int data)
{
	Node *nodeToDelete = list->head;
	Node *prevNode = nullptr;
	while (nodeToDelete->data != data && nodeToDelete->next != nullptr)
	{
		prevNode = nodeToDelete;
		nodeToDelete = nodeToDelete->next;
	}
	if (nodeToDelete->next == nullptr && nodeToDelete->data != data)
	{
		printf("Value is not found\n");
		return;
	}
	if (nodeToDelete != list->head)
	{
		prevNode->next = nodeToDelete->next;
	}
	else
	{
		list->head = nodeToDelete->next;
	}
	delete nodeToDelete;
	--list->length;
}

void deleteList(List *list)
{
	for (int i = 0; i < list->length; ++i)
	{
		Node *nodeToDelete = list->head;
		list->head = nodeToDelete->next;
		delete nodeToDelete;
	}
}

void printList(List *list)
{
	if (isEmpty(list))
	{
		printf("The list is empty\n");
		return;
	}
	Node *nodeToPrint = list->head;
	while (nodeToPrint != nullptr)
	{
		printf("%d ", nodeToPrint->data);
		nodeToPrint = nodeToPrint->next;
	}
	printf("\n");
}

bool test()
{
	List *testList = createList();
	addNode(testList, 4);
	addNode(testList, -5);
	addNode(testList, 8);
	addNode(testList, 0);
	addNode(testList, 1);
	deleteNode(testList, 4);
	Node *nodeForTest = testList->head;
	Node *nodeForTestNext = nodeForTest->next;
	for (int i = 0; i < testList->length - 1; ++i)
	{
		if (nodeForTest->data > nodeForTestNext->data || nodeForTest->data == 4)
		{
			deleteList(testList);
			return false;
		}
		nodeForTest = nodeForTest->next;
		nodeForTestNext = nodeForTestNext->next;
	}
	deleteList(testList);
	return true;
}

File: sem2/hw2/hw2.2/hw2.2/List.cs
using System;

namespace hw2._2
{
    /// <summary>
    /// Singly linked list, a linear collection of integer values.
    /// </summary>
    public class List : IList
    {
        private class Node
        {
            public int Data { get; set; }
            public Node Next { get; set; }

            public Node(int data, Node next)
            {
                Data = data;
                Next = next;
            }
        }

        private Node head;

        /// <summary>
        /// Number of elements in list.
        /// </summary>
        public int Size { get; private set; }

        public List()
        {
        }

        /// <summary>
        /// Checks if the list has no elements in it.
        /// </summary>
        /// <returns>True if the list is empty, else false.</returns>
        public bool IsEmpty() => head == null;

        private bool IsCorrectPositionToAdd(int position) => position > 0 && position <= Size + 1;

        private bool IsCorrectPosition(int position) => position > 0 && position <= Size;

        private Node GetNode(int position)
        {
            Node current = head;
            for (var i = 1; i < position; ++i)
            {
                current = current.Next;
            }
            return current;
        }

        /// <summary>
        /// Adds new element in list by its position.
        /// </summary>
        /// <param name="position">Position of the element to add.</param>
        /// <param name="data">Element to add.</param>
        /// <returns>True if the element is added, false if the given position is incorrect.</returns>
        public bool AddElement(int position, int data)
        {
            if (!IsCorrectPositionToAdd(position))
            {
                return false;
            }

            var newNode = new Node(data, null);

            if (position == 1)
            {
                newNode.Next = head;
                head = newNode;
            }
            else
            {
                var current = GetNode(position - 1);
                newNode.Next = current.Next;
                current.Next = newNode;
            }
            ++Size;
            return true;
        }

        /// <summary>
        /// Deletes an element from the list by its position.
        /// </summary>
        /// <param name="position">Position of the element to delete.</param>
        /// <returns>True if the element is deleted, false if the given position is incorrect.</returns>
        public bool DeleteElementByPosition(int position)
        {
            if (!IsCorrectPosition(position))
            {
                return false;
            }

            if (position == 1)
            {
                head = head.Next;
            }
            else
            {
                Node current = GetNode(position - 1);
                current.Next = current.Next.Next;
            }
            --Size;
            return true;
        }

        /// <summary>
        /// Deletes an element from the list by its data.
        /// </summary>
        /// <param name="data">Data of the element to delete.</param>
        public void DeleteElementByData(int data)
        {
            if (head.Data == data)
            {
                head = head.Next;
                --Size;
                return;
            }

            Node previous = head;
            Node current = head.Next;

            for (var i = 0; i < Size; ++i)
            {
                if (current.Data == data)
                {
                    previous.Next = current.Next;
                    --Size;
                    return;
                }
                previous = current;
                current = current.Next;
            }
        }

        /// <summary>
        /// Gets data of the list element by its position.
        /// </summary>
        /// <param name="position">Position of the element.</param>
        /// <returns>Data of the element or ' ' if the given position is incorrect.</returns>
        public int GetData(int position)
        {
            if (!IsCorrectPosition(position))
            {
                return ' ';
            }
            Node node = GetNode(position);
            return node.Data;
        }

        /// <summary>
        /// Sets data of the list element by its position.
        /// </summary>
        /// <param name="position">Position of the element.</param>
        /// <param name="data">Data of the element.</param>
        /// <returns>True if data is set successfully, false if the given position is incorrect.</returns>
        public bool SetData(int position, int data)
        {
            if (!IsCorrectPosition(position))
            {
                return false;
            }

            Node node = GetNode(position);
            node.Data = data;
            return true;
        }

        /// <summary>
        /// Prints the list.
        /// </summary>
        public void PrintList()
        {
            Node current = head;
            while (current != null)
            {
                Console.Write($"{current.Data} ");
                current = current.Next;
            }
        }

        /// <summary>
        /// Checks if the list element exists.
        /// </summary>
        /// <param name="data">Element to check.</param>
        /// <returns>True if the list element is found, else false</returns>
        public bool Exists(int data)
        {
            Node current = head;
            for (var i = 0; i < Size; ++i)
            {
                if (current.Data == data)
                {
                    return true;
                }
                current = current.Next;
            }
            return false;
        }
    }
}